Update deployment walkthrough to allow configurable domain (#218)
mspnp/aks-secure-baseline: 09-workload.md

> :book: The Contoso app team is about to conclude this journey, but they need an app to test their new infrastructure. For this task they've picked out the venerable [ASP.NET Core Docker sample web app](https://github.com/dotnet/dotnet-docker/tree/master/samples/aspnetapp).

> 1. Customize the host name of the Ingress resource to match your custom domain
> 
>    > An example file is provided that will be used later with the built-in kubectl customization tool, Kustomize. This lets you patch the default Ingress resource definition with your own custom values.
> 

Reviewer: Part of a kustomization/workflow simplification suggestion (part 4 of 4):

```suggestion
1. Customize the host name of the Ingress resource to match your custom domain. _(You can skip this step if domain was left as contoso.com.)_

```

   ```bash

>    sed "s/<domain-name>/${DOMAIN_NAME}/g" workload/aspnetapp-ingress-patch.yaml.example > workload/aspnetapp-ingress-patch.yaml

Reviewer: Part of a kustomization/workflow simplification suggestion (part 4 of 4):

1. Customize the host name of the Ingress resource to match your custom domain. _(Okay to skip this step if domain was left as contoso.com.)_

```suggestion
   sed -i "s/contoso.com/${DOMAIN_NAME}/" workload/aspnetapp-ingress-patch.yaml
```

   ```

1. Deploy the Ingress resource

   ```bash
   cat <<EOF | kubectl create -f -
   apiVersion: networking.k8s.io/v1
   kind: Ingress
   metadata:
     name: aspnetapp-ingress
     namespace: a0008
     annotations:
       kubernetes.io/ingress.allow-http: "false"
       # defines controller implementing this ingress resource: https://docs.microsoft.com/en-us/azure/dev-spaces/how-to/ingress-https-traefik
       # ingress.class annotation is being deprecated in Kubernetes 1.18: https://kubernetes.io/docs/concepts/services-networking/ingress/#deprecated-annotation
       # For backwards compatibility, when this annotation is set, precedence is given over the new field ingressClassName under spec.
       kubernetes.io/ingress.class: traefik-internal
       traefik.ingress.kubernetes.io/router.entrypoints: websecure
       traefik.ingress.kubernetes.io/router.tls: "true"
       traefik.ingress.kubernetes.io/router.tls.options: default
       traefik.ingress.kubernetes.io/router.middlewares: app-gateway-snet@file, gzip-compress@file
   spec:
     # ingressClassName: "traefik-internal"
     tls:
     - hosts:

>          - bu0001a0008-00.aks-ingress.$DOMAIN_NAME

Reviewer: ```suggestion
         - bu0001a0008-00.aks-ingress.${DOMAIN_NAME}
```

           # it is possible to opt for certificate management strategy with dedicated
           # certificates for each TLS SNI route.
           # In this Rereference Implementation for the sake of simplicity we use a
           # wildcard default certificate added at Ingress Controller configuration level which is *.example.com

>            # secretName: <bu0001a0008-00-example-com-tls-secret>

Reviewer: This is set up as the default in Traefik's config, https://github.com/mspnp/aks-secure-baseline/blob/b3319376ac58fe28ad5244be2776bbf84633b47a/workload/traefik.yaml#L295 -- which has the other name hardcoded in there.  I'm thinking that secret name (KV and CSI KV Driver reference) should be genericized as well as part of this PR.

Reviewer: Looks like you've got them all updated, nice!

     rules:

>      - host: bu0001a0008-00.aks-ingress.$DOMAIN_NAME

Reviewer: ```suggestion
     - host: bu0001a0008-00.aks-ingress.${DOMAIN_NAME}
```

       http:
         paths:
         - path: /
           pathType: Prefix
           backend:
             service:
               name: aspnetapp-service
               port:
                 number: 80
   EOF
   ```

> 

Reviewer: We try to minimize inline yaml creation like this, (I know we have it in another spot here as well -- but trying to prevent more).  Any thoughts on maybe adding a Kustomization to the workload directory with a strategicMerge patch, then instead of this, we could do a quick `sed` command to replace a placeholder in that file, and instead of applying the workload with `-f` we could just `-k` that directory. There is a lot of "comment noise" in this file, which I think is best left in a file vs echo'd directly into a kubectl apply and to be lost.  Thoughts?

Author: I like that approach. I'll work on making the changes.

Reviewer: Good start on this, I'm going to resolve this comment here, because my follow-on review has a slight tweak to the implementation, but the end result is the same.

   kubectl run curl -n a0008 -i --tty --rm --image=mcr.microsoft.com/azure-cli --limits='cpu=200m,memory=128Mi'
   
   # From within the open shell

>    DOMAIN_NAME="contoso.com"

Reviewer: ```suggestion
   DOMAIN_NAME="contoso.com" # <-- Change to your custom domain value if a different one was used
```
